scalafix-core/shared/src/main/scala/scalafix/internal/util/TypeSyntax.scala

```diff
@@ -1,138 +1,452 @@
+import scala.meta.internal.semanticdb3.SymbolInformation.{Property => p}
+import scala.meta.internal.semanticdb3.Type.{Tag => t}
+import TypeSyntax._
+import scalapb.GeneratedMessage
 
 object TypeSyntax {
-  def prettify(tpe: Type, ctx: RuleCtx, shortenNames: Boolean)(
-      implicit index: SemanticdbIndex): (Type, Patch) = {
+  def prettify(
+      stpe: s.Type,
+      ctx: RuleCtx,
+      shortenNames: Boolean,
+      pos: Position)(implicit index: SemanticdbIndex): Option[(Type, Patch)] = {
+    try {
+      val table = index.asInstanceOf[EagerInMemorySemanticdbIndex]
+      Some(new TypeSyntax(stpe, ctx, shortenNames, pos, table).prettify())
+    } catch {
+      case NonFatal(e) =>
+        val msg = new StringWriter()
+        e.printStackTrace(new PrintWriter(msg))
+        ctx.config.reporter.error(msg.toString, pos)
+        None
+    }
+  }
 
-    val functionN: SymbolMatcher = SymbolMatcher.exact(
-      0.to(22).map(i => Symbol(s"_root_.scala.Function$i#")): _*
-    )
-    val tupleN: SymbolMatcher = SymbolMatcher.exact(
-      1.to(22).map(i => Symbol(s"_root_.scala.Tuple$i#")): _*
-    )
+  implicit class XtensionSymbolInformationProperties(
+      info: s.SymbolInformation) {
+    private def test(property: s.SymbolInformation.Property): Boolean =
+      (info.properties & property.value) != 0
+    def isVal: Boolean = test(p.VAL)
+    def isVar: Boolean = test(p.VAR)
+    def isVarSetter: Boolean =
+      isVar && info.name.endsWith("_=")
+  }
 
-    def isStable(symbol: Symbol): Boolean = {
-      def loop(symbol: Symbol): Boolean = symbol match {
-        case Symbol.None => true
-        case Symbol.Global(owner, Signature.Term(_)) => loop(owner)
-        case _ => false
-      }
-      symbol match {
-        case Symbol.Global(owner, Signature.Term(_) | Signature.Type(_)) =>
-          loop(owner)
+  object T {
+    abstract class TypeRefExtractor(sym: String) {
+      def matches(tpe: s.Type): Boolean = tpe.tag match {
+        case t.TYPE_REF =>
+          tpe.typeRef.exists(_.symbol == sym)
+        case t.WITH_TYPE =>
+          tpe.withType.exists { x =>
+            x.types.lengthCompare(1) == 0 &&
+            unapply(x.types.head)
+          }
         case _ =>
           false
       }
+      def unapply(tpe: s.Type): Boolean = matches(tpe)
+    }
+    object AnyRef extends TypeRefExtractor("scala.AnyRef#")
+    object Nothing extends TypeRefExtractor("scala.Nothing#")
+    object Any extends TypeRefExtractor("scala.Any#")
+  }
+
+  def isFunctionN(symbol: String): Boolean = {
+    symbol.startsWith("scala.Function") &&
+    symbol.endsWith("#")
+  }
+
+  def isTupleN(symbol: String): Boolean = {
+    symbol.startsWith("scala.Tuple") &&
+    symbol.endsWith("#")
+  }
+
+  object FunctionN {
+    def unapply(symbol: String): Boolean = isFunctionN(symbol)
+  }
+  object TupleN {
+    def unapply(symbol: String): Boolean = isTupleN(symbol)
+  }
+}
+
+class TypeSyntax private (
```

Comment:
Would you ultimately want parts of this code to end up in Metap? https://github.com/scalameta/scalameta/issues/1479

Reply:
I'd like to rewrite this file from scratch first before considering that. I am not happy with the code quality here, but I am more eager to upgrade to scalameta v3.7 and this PR is already doing too much.